includes/payment-methods/class-upe-payment-method.php

```diff
@@ -0,0 +1,128 @@
+<?php
+/**
+ * Abstract UPE Payment Method class
+ *
+ * Handles general functionality for UPE payment methods
+ *
+ * @package WCPay\Payment_Methods
+ */
+
+namespace WCPay\Payment_Methods;
+
+use WP_User;
+use WC_Payments_Token_Service;
+use WC_Payment_Token_CC;
+use WC_Payment_Token_WCPay_SEPA;
+use WC_Subscriptions;
+use WC_Subscriptions_Cart;
+
+/**
+ * Extendable abstract class for payment methods.
+ */
+abstract class UPE_Payment_Method {
+
+	/**
+	 * Stripe key name
+	 *
+	 * @var string
+	 */
+	protected $stripe_id;
+
+	/**
+	 * Display title
+	 *
+	 * @var string
+	 */
+	protected $title;
+
+	/**
+	 * Can payment method be saved or reused?
+	 *
+	 * @var bool
+	 */
+	protected $is_reusable;
+
+	/**
+	 * Instance of WC Payments Token Service to save payment method
+	 *
+	 * @var WC_Payments_Token_Service
+	 */
+	protected $token_service;
+
+	/**
+	 * Create instance of payment method
+	 *
+	 * @param WC_Payments_Token_Service $token_service Instance of WC_Payments_Token_Service.
+	 */
+	public function __construct( $token_service ) {
+		$this->token_service = $token_service;
+	}
+
+	/**
+	 * Returns payment method ID
+	 *
+	 * @return string
+	 */
+	public function get_id() {
+		return $this->stripe_id;
+	}
+
+	/**
+	 * Returns payment method title
+	 *
+	 * @param array|bool $payment_details Optional payment details from charge object.
+	 *
+	 * @return string
+	 */
+	public function get_title( $payment_details = false ) {
+		return $this->title;
+	}
+
+	/**
+	 * Returns boolean dependent on whether payment method
+	 * can be used at checkout
+	 *
+	 * @return bool
+	 */
+	public function is_enabled_at_checkout() {
+		if ( $this->is_subscription_item_in_cart() ) {
+			return $this->is_reusable();
+		}
+		return true;
+	}
+
+	/**
+	 * Returns boolean dependent on whether payment method
+	 * will support saved payments/subscription payments
+	 *
+	 * @return bool
+	 */
+	public function is_reusable() {
+		return $this->is_reusable;
+	}
+
+	/**
+	 * Add payment method to user and return WC payment token
+	 *
+	 * @param WP_User $user User to get payment token from.
+	 * @param string  $payment_method_id Stripe payment method ID string.
+	 *
+	 * @return WC_Payment_Token_CC|WC_Payment_Token_WCPay_SEPA WC object for payment token.
+	 */
+	public function get_payment_token_for_user( $user, $payment_method_id ) {
+		return $this->token_service->add_payment_method_to_user( $payment_method_id, $user );
+	}
+
+	/**
+	 * Returns boolean on whether current WC_Cart or WC_Subscriptions_Cart
+	 * contains a subscription or subscription renewal item
+	 *
+	 * @return bool
+	 */
+	public function is_subscription_item_in_cart() {
+		if ( class_exists( 'WC_Subscriptions' ) && version_compare( WC_Subscriptions::$version, '2.2.0', '>=' ) ) {
```

Comment:
```suggestion
		if ( $this->is_subscriptions_enabled() ) {
```
We could use the function from the trait.

Reply:
@brezocordero I borrowed that from the trait, but unfortunately I can't use the trait directly in this class without implementing the other abstract functions from the trait, which don't really have any relevance to this class. 

I don't like repeating this logic in two places, so let me know if you have any better ideas to resolve this! Otherwise I'll sweep it under the rug and continue with the merge.

Reply:
Ignore my comment, I thought it was inside of `includes/payment-methods/class-upe-payment-gateway.php` 
Good call to sweep it.

Reply:
Would it be possible to call that function through the parent? If not, I'd say just duplicate this check for right now to keep things moving.

Reply:
`UPE_Payment_Method` is a separate abstract base class that doesn't inherit from the UPE gateway, so doesn't rub shoulders with any of that code. I don't think there's enough shared functionality between the gateways and the methods to need to tie them together, plus that keeps the payment methods lightweight as we add new ones. 

What I could do is split the Subscriptions trait into two traits--maybe one with the boolean checks and one with the gateway functions--then use the smaller trait inside the bigger one and `UPE_Payment_Method` and use the bigger one inside the gateway, as it is right now. Let me know if you think that's a good idea and I could fix that up quickly and merge.